core/iwasm/libraries/lib-wasi-threads/test/spawn_stress_test.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2023 Amazon.com Inc. or its affiliates. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#ifndef __wasi__
+#error This example only compiles to WASM/WASI target
+#endif
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <assert.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <math.h>
+#include <pthread.h>
+#include <string.h>
+
+
+enum CONSTANTS {
+    NUM_ITER = 100000,
+    NUM_RETRY = 5,
+    SECOND = 1000 * 1000 * 1000, /* 1 second */
+    TIMEOUT = 50LL * SECOND,
+    MAX_NUM_THREADS = 8,
+};
+
+unsigned int factorised_number = 2;
+unsigned int prime_numbers = 0;
+double percentage = 0.1;
+pthread_t threads[MAX_NUM_THREADS];
+unsigned int args[MAX_NUM_THREADS];
+
+bool is_prime(unsigned int num) {
+    for (unsigned int  i = 2; i <= (unsigned int)(sqrt(num)); ++i) {
+        if (num % i == 0) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void* check_if_prime(void* value) {
+    unsigned int* num = (unsigned int*)(value);
+    __builtin_wasm_memory_atomic_wait32(NULL, 0, 0.01 * SECOND);
+    if (is_prime(*num)) {
+        __atomic_fetch_add(&prime_numbers, 1, __ATOMIC_SEQ_CST);
+    }
+    return NULL;
+}
+
+unsigned int validate() {
+    unsigned int counter = 0;
+    for (unsigned int i = 2; i <= factorised_number; ++i) {
+        counter += is_prime(i);
+    }
+
+    return counter;
+}
+
+int
+main(int argc, char **argv)
+{
+    for (; factorised_number < NUM_ITER; ++factorised_number) {
+        unsigned int thread_num = factorised_number % MAX_NUM_THREADS;
+        args[thread_num] = factorised_number;
+        if (threads[thread_num] != 0) {
+            assert (pthread_join(threads[thread_num], NULL) == 0);
+        }
+        assert (pthread_create(&threads[thread_num], NULL, &check_if_prime, &args[thread_num]) == 0 && "Thread creation should succeed");
```

Comment:
Because both `pthread_create` and `pthread_join` are implemented in userspace, it's possible that a new native thread will start before the other finishes. I'd suggest adding retry mechanism for `pthread_create` and try a few times on the `EAGAIN` error (with some interval) before failing on assertion.

Reply:
yeah, that makes sense. Thank you